### DISCUSSION TITLE
Making Sequoia More Transparent

### DISCUSSION THREAD

**Discussion by lebrice** [lebrice/Sequoia#139] [General]
I have a feeling that Sequoia isn't as clear to new users as it could be.

This issue is meant primarily as a way for me to list out some of the design challenges I've been facing, as well as to share them with you all. I would love to hear other perspectives on this. I might update this thread as new ideas / suggestions pop up.

Goals / Things to address:
- New users should get a clear understanding of the "main loop" of any Setting in Sequoia, by just opening the corresponding source file.
- It should become evident to users why polymorphism/inheritance makes sense when talking about evaluation "Settings" in ML. It shouldn't feel over-engineered.
- It should be more clear what is gained from making a distinction (in code) between the roles of the `Setting` vs the `Method`, as well as where abstractions like `configure()`, `fit`, `get_actions` come from, and why they are useful.

- Need to decide what kind of inheritance hierarchy (tree) we want to construct long-term:
    - "pure inheritance":
        - simplified tree, most logic common to RL/SL goes in the root node.
        - No real way to create a Method that works on both sides of the Tree, but only after a given layer, for example a Method that works in both RL and SL, but requires task labels at test time.
        
    - "pure composition":
        - A Setting is just the sum of its assumptions. To create a new Setting, we create a new class that inherits from all the desired assumptions.
        (e.g. `PassiveAssumption`  + `IncrementalAssumption` + `TaskAwareAssumption` --> `TaskIncrementalSLSetting`). 
        - Each assumption has to contain logic that is absolutely independent from other assumptions.
        - If we fail to keep assumptions absolutely independent from each other, I have a feeling that the entire tree might collapse, as we might have to deal with the ramifications of polynomial (or is it exponential?) interactions between all families/types of assumptions.
        For example this `IncrementalAssumption` might do need to do something slightly differently when creating the environments from each task depending on if the concrete setting is a passive (SL) or active (RL) Setting.
 
    - **"Hybrid"** (closer to current setup):
        - There is a simple hierarchy going from Setting at the top, downwards on either the RL or SL side until the "IID" settings are reached.
        - There is also a set of "Assumptions" (mixins) that hold the code that is common to both sides of the tree.
        - All settings use the same "main_loop" defined in the `IncrementalSetting` (meant to act as the `IncrementalAssumption` from above), which doesn't really make sense:
            - `ContinualRLSetting`, for instance, should be more 'general' than the IncrementalSetting, since it doesn't have task boundaries, and the environment shifts continuously over time. However, it currently subclasses IncrementalSetting, which is confusing. (Might want to make a new Issue/PR for this)

Potential TODOs / things to address:
- [ ] Simplify the main_loop as much as possible
- [ ] Add the pseudocode of the main loop, or the main loop itself (hopefully not just copy-pasted) to all Settings.  
- [ ] Add an example in the style of Pytorch Lightning (Before / after side-by-side comparison) showing how to run a "simple" CL evaluation loop (e.g. Class-Incremental MNIST), without any abstraction or inheritance or anything. Just one script that does everything, from start to finish, using just Continuum. Then try to do a slightly tweaked version of this setup or Method. Then change the Method, etc, until it becomes clear why the separation of concerns makes sense, and why polymorphism is a natural fit for this kind of growing complexity problem.
- `class_incremental_setting.py` is getting too long!
    - [ ] Stuff related to creating dataloaders from gym environments (specifically) should be placed in the "active" package, rather than all of it being in `class_incremental_setting.py`
- `continual_rl_setting.py` is getting too long!
    - [ ] Stuff related to creating gym environments from datasets (specifically) should be placed in the "passive" package.

Let me know what you think!

**Comment by optimass** [comment]
My mental model is "pure composition" but for the reasons you listed, it might lead to the brittlest implementation of Sequoia. That being said I guess it's the one that (in theory at least) would be the easiest to graft a new assumption to add settings.

"Pure inheritance" is aligned with our mathematical definition and makes a lot of sense theoretically, but my guess is grafting new assumptions that are not at the bottom of the tree would be a nightmare. that being said, can you get more general than RL? 

Codewise I guess "Hybrid" makes sense. I think we could view Sequoia v1 as an MVP and decide later if we need a v2 depending on the demand for Sequoia. 

On the TODOs:

~~I would wait before simplifying the main_lopp + add pseudocode: I think that the average users don't care how the sausage is made, as long as it's good. If there is a demand for that, we can reconvene.~~

Ok actually maybe main_loop pseudocode is good, as it's less about understanding the backend and more about understanding the settings.

love the Pytorch Lightning example. Picture = 1000 words. This solves bullet point 3.
